Plans: skip plans link for free plan

rralian opened the pull request:
This PR implemented an abtest for an updated plans page in the NUX flow that no longer presents "free" as a plan option, but rather as what you will get if you skip the plans step. This idea has been tested before with mixed results, but the addition of the "personal plan" otherwise forces us to present a four-column page, and we think presenting the plans in a three-column layout will perform better for selection and purchases.

## Visual Changes
We are removing the free plan and replacing it with a "skip for now" navigation button above and below the plan options. This is the same button and wording we present on the themes step for a user to skip themes selection. Also notice that the action buttons are updated to say "Select Personal", "Select Premium", and "Select Business" instead of "Upgrade Now". This is because "upgrade" as a verb doesn't make as much sense if the default action is to select a paid plan.

Here is the updated plans step.
![plans step](https://cloudup.com/c3kSZpcQaOs+)

And here is the accompanying plans/compare page
![plans compare](https://cloudup.com/cisK-7EO3hI+)

## Testing
As part of this PR we are re-enabling the personal plan, setting it to 100%. This change will only be presented when the "personal plan" is enabled. So if you have the personal plan disabled via localStorage (i.e. set to "hide"), you should not see these changes.

Go through the NUX flow with abtest values `personalPlan = show` and `skipPlansLinkForFree = skipPlansForFree` and ensure you see the plans and plans/compare pages as described. Go directly to these pages for a free plan in your sidebar and you should _not_ see these changes... non-NUX plan pages should be unaffected. Now disable either the personal plan or the skipPlans tests and go through the NUX again and you should see the plans pages presented in a four-column layout with the free plan on the left.

Also, with an incognito window, going through the NUX, ensure you don't see the `calypso_abtest_start` event until you get to the plans-selection step of the NUX.

![calypso_abtest_start](https://cloudup.com/cA9Z92cPyJi+)

/cc @rodrigoi @apeatling @gwwar @meremagee 

Test live: https://calypso.live/?branch=update/skip-plans-link-for-free

shaunandrews commented:
I like the concept of this test a lot, but I'm a little worried that the skip link is way too hidden. Perhaps thats the point, as this will likely "force" people to click one of the 3 bright blue buttons?

From a recent run of user tests, I noticed that many people don't see the buttons at all as they're hidden "below the fold." Instead of looking for the buttons, users end up clicking on the plan name to continue — they never scroll down.

Without an obvious way to move forward for free (and "FREE" being a big part of at least one homepage variation) I think a lot of people will simply drop out without ever noticing that they can skip forward.

rralian commented:
Good points @shaunandrews, thank you. We are working on a more ambitious redesign of the plans page that includes moving the action buttons to the top. But we wanted to test this change separately. Definitely open to feedback on wording and placement for the skip button.

Part of the reason I chose this approach was because it's what we already had implemented for a "skip" button and I wasn't sure I should add another concept. I'm not super-fond of the "skip for now" wording actually, as it presents the action as a decision that can be deferred, and I would rather present it as a conscious choice _not_ to get our fancy features. So something like "continue with the free plan without these premium features", but without all that jumble of words. Suggestions?

shaunandrews commented:
> We are working on a more ambitious redesign of the plans page that includes moving the action buttons to the top.

Yup. I've been watching. :) I get the want to test this out quickly. Makes sense.

> I'm not super-fond of the "skip for now" wording actually

I think thats ok for now. Its consistent with the themes step, and even with the position being less than ideal, I'm happier with the consistency for now.

> Suggestions?

The first idea that jumps in my head is to keep the "skip for now" button thats there, but also change the subheading on the page to mention the ability to continue for free:

![screen shot 2016-07-12 at 10 32 17 am](https://cloud.githubusercontent.com/assets/191598/16770723/f1c3a8ec-481b-11e6-84ba-1e956fd65c99.png)

rralian commented:
Could we tweak that so it's presented as a choice, rather than deferring a decision?
> Don't need these features? Continue with a <a href="#">free basic site</a>.

shaunandrews commented:
Fur sure. I think that works just as well.

apeatling commented:
Would it be tough to just duplicate the skip for now link at the top and align it to the right, like in my mockup? I think that will give us the most honest results based on what we expect to do with the new design.

shaunandrews commented:
> like in my mockup

What mockup?

rralian commented:
@apeatling I see. I was going by the recap in our slack channel yesterday.
> so it’d be personal instead of free here + skip link on the bottom for v1

IMHO the "skip for now" language is one of the reasons the previous test failed. It gives the action no cost. I'd rather the user feel at least some amount of [FOMO](https://en.wikipedia.org/wiki/Fear_of_missing_out) regarding our paid features. But if y'all are in agreement, I can work that up.

apeatling commented:
I am getting stuck on why we would test something that we know is bad though. We know that dropping the skip link below the fold is bad, because most people will miss it, and it's not something we would ever do in an updated design.

rralian commented:
Thanks for the feedback @apeatling and @shaunandrews. I've updated to reflect design changes we discussed in slack and I've updated the screenshot in the description at the top. This PR could definitely use some heavy testing to make sure I am _not_ introducing changes to the plans and plans/compare pages as available from the sidebar and that I am _not_ introducing changes when this abtest is disabled.

rralian commented:
Some refactoring in light of #6725 coming through sometime after this one. I also believe I had some issues in my logic for testing whether this test was enabled that should be cleared up now. And last, I'm only enabling the `personalPlan` and `skipPlansLinkForFree` features in development and wpcalypso, so we can get more solid testing before we launch this to production. /cc @rodrigoi

dmsnell commented on client/components/plans/plan-actions/index.jsx:
```diff
@@ -225,7 +234,9 @@ const PlanActions = React.createClass( {
-				<a href={ link } className="button plan-actions__upgrade-button">{ this.translate( 'Manage Plan', { context: 'Link to current plan from /plans/' } ) }</a>
+				<a href={ link } className="button plan-actions__upgrade-button">{
+						this.translate( 'Manage Plan', { context: 'Link to current plan from /plans/' } )
+				}</a>
```
are doing this with the brackets now? why not…

```js
{ this.translate( … ) }
```

gwwar replied:
Do you mean providing the `<a>` tag as a component? In this case since we don't have mixed html/strings here I think the current usage is ok.

dmsnell replied:
it was about the formatting/style of putting the brackets with the `<a />` tag and not with `this.translate()`

dmsnell commented on client/signup/step-wrapper/index.jsx:
```diff
@@ -65,26 +74,62 @@ export default React.createClass( {
 
+	recordComparePlansClick() {
+		analytics.ga.recordEvent( 'Upgrades', 'Clicked Compare Plans Link' );
+	},
```
Can we stick with the new Redux analytics middleware?

```js
import { recordGoogleEvent } from 'state/analytics/actions';

recordComparePlansClick() {
	this.props.trackCompareClick();
}

const mapDispatchToProps = {
	trackCompareClick: recordGoogleEvent( 'Upgrades', 'Clicked Compare Plans Link' )
}
```

dmsnell commented on client/signup/step-wrapper/index.jsx:
```diff
@@ -65,26 +74,62 @@ export default React.createClass( {
+		return (
+			<a href={ url } className="step-wrapper__compare-plans" onClick={ this.recordComparePlansClick }>
+				<Gridicon icon="clipboard" size={ 18 } />
+				{ compareString }
```
We planning on doing more with this? You used to suggest that I inline inside of `render()` as much as possible :wink:

rralian replied:
I think there's enough logic here to be useful to extract from the main render. The different buttons pop into and out of existence depending on your step so it's easier to consider them on their own. But consider yourself vindicated.

dmsnell replied:
Well, I wasn't speaking of the overall picture, just `compareString`

rralian replied:
Oh, do you mean how compareString is generated earlier and then added here? I suppose. This could have just used `this.translate()` inside the brackets. Truth be told, I probably would have done that, but I believe I copied this from somewhere else and then modified it for my own use. 😊

dmsnell commented on client/signup/steps/plans/index.jsx:
```diff
@@ -138,6 +141,7 @@ module.exports = React.createClass( {
 		return (
 			<StepWrapper
 				flowName={ this.props.flowName }
+				goToNextStep={ isSkipPlansTestEnabled() ? this.onSelectPlan : null }
```
Looks like `goToNextStep` is looking for a function. Should we send it `noop` instead of `null`? Further, if we simply don't want to send it, maybe `undefined` would be more appropriate.

gwwar replied:
+1 for setting this to `undefined` so the component can set a default if it has one.

apeatling commented:
Not quite sure what's going on with the buttons here, they are all misaligned. I think this may also be the case for prod though, although I can't see the personal plan there.

![screen shot 2016-07-12 at 3 48 51 pm](https://cloud.githubusercontent.com/assets/1464705/16786258/58d8c3e2-4848-11e6-932e-cdb18559b6eb.png)

apeatling commented:
Looks like I broke it with my changes, fixing. :)

apeatling commented:
If I disable the personal plan test, but leave skipfree test enabled, I lose the plans comparison link under the title in the NUX step. Not sure if this is an issue or not, if both are disabled, then I see it.

rralian commented:
> If I disable the personal plan test, but leave skipfree test enabled, I lose the plans comparison link under the title in the NUX step. Not sure if this is an issue or not, if both are disabled, then I see it.

I don't _believe_ this is a big issue, but I can try to address it. It's my intent that the skip-free will only ever be displayed when personalPlan is also enabled.

apeatling commented:
All looks fine both in the NUX and plans page, tests full off and on other than that.

gwwar commented on client/lib/plans/index.js:
```diff
@@ -262,3 +262,10 @@ export function applyTestFiltersToPlansList( planName ) {
+	return isEnabled( 'plans/skip-plans-for-free' ) &&
+		isEnabled( 'plans/personal-plan' ) &&
+		abtest( 'personalPlan' ) === 'show' &&
+		abtest( 'skipPlansLinkForFree' ) === 'skipPlansForFree';
```
If a user already has the personalPlan variant set, does 100% force them to show? Otherwise folks might have `skipPlansLinkForFree` and `personalPlan` to hide which may make the results a bit more difficult to look at.

rralian replied:
It does not force existing users to 'show', so users who already have the variant set to 'hide' would still see 'hide'. That wouldn't impact the results of a test because I would only be testing against new users. But it could lead to some UX weirdness like @apeatling pointed out, so I am going to go through and address that. The abtest will be cleared out once @rodrigoi goes through with #6725 so this code will be somewhat short-lived.

gwwar commented:
👍 Tested through the major flows, and I tried a some other currencies. The current code comments don't look like show stoppers. I'd say go ahead and 🚢 this tomorrow morning if no one else finds any blockers.

lamosty commented on client/signup/step-wrapper/index.jsx:
```diff
@@ -138,7 +138,7 @@ const StepWrapper = React.createClass( {
 
 const mapDispatchToProps = dispatch => ( {
-	trackCompareClick: dispatch( recordGoogleEvent( 'Upgrades', 'Clicked Compare Plans Link' ) )
+	trackCompareClick: () => dispatch( recordGoogleEvent( 'Upgrades', 'Clicked Compare Plans Link' ) )
```
Why? I've just watched a Redux tutorial and they don't return the dispatch.

rralian commented:
> If I disable the personal plan test, but leave skipfree test enabled, I lose the plans comparison link under the title in the NUX step. Not sure if this is an issue or not, if both are disabled, then I see it.

You were right to point this out, thanks. I've addressed it correctly now. Every UX change now checks against the `isSkipPlansTestEnabled()` utility function, which checks for _both_ the `personalPlan` and `skipPlansLinkForFree` features being enabled in config and abtest before changing anything.

One little thing @apeatling, your css tweaks are getting applied to both variants. I think they're fine, but want to make sure you're aware of that and that it was your intention.
